subprojects/core/src/main/java/org/gradle/api/internal/tasks/TaskOptionSupplier.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.gradle.api.internal.tasks.options.BooleanOptionElement;
+import org.gradle.api.internal.tasks.options.BuiltInOptionElement;
+import org.gradle.api.internal.tasks.options.InstanceOptionDescriptor;
+import org.gradle.api.internal.tasks.options.OptionDescriptor;
+import org.gradle.api.internal.tasks.options.OptionElement;
+import org.gradle.api.internal.tasks.options.OptionReader;
+import org.gradle.api.specs.Specs;
+import org.gradle.util.internal.CollectionUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * This class is responsible for adding built-in task options and
+ * {@link BooleanOptionElement boolean opposite options} to the options
+ * provided by the {@link OptionReader}.
+ */
+public class TaskOptionSupplier {
+    private static final Logger LOGGER = LoggerFactory.getLogger(TaskOptionSupplier.class);
+    private final Object target;
+    private final OptionReader optionReader;
+    private final List<OptionDescriptor> oppositeOptionPairs = new LinkedList<>();
```

Comment:
🎨 This new state here makes this object non-reusable and would cause suprise if it is reused. As we discussed I think instead of the two public functions which require holding this state we should have a single public function that returns an object that contains both all-options and mutually-exclusive-options.